domain/application/service/provider.go

```diff
@@ -1099,6 +1099,7 @@ func makeCreateApplicationArgs(
 	return application.BaseAddApplicationArg{
 		Charm:             ch,
 		CharmDownloadInfo: args.DownloadInfo,
+		Constraints:       constraints.DecodeConstraints(args.Constraints),
```

Comment:
There is a separate but related bug before we reach this point. `makeCAASApplicationArg` and `makeIAASApplicationArg` receive the application constraint and both call `resolveApplicationConstraints` to merge together the app and model constraints. But the merged result is not passed down into `makeApplicationArg` which eventually reach here.

Reply:
I can fix this in a follow-up if you'd prefer to keep this PR as-is.

Reply:
I agree with @kian99 - happy to do in a follow up.

Reply:
Done!, here's the QA steps for this:
```
juju set-model-constraints mem=3072
juju deploy juju-qa-dummy-source --constraints cpu=42 
juju constraints dummy-source
cores=42 mem=3072M
```